### DISCUSSION TITLE
Develop customized or third pkg for CUE

### DISCUSSION THREAD

**Discussion by yangzhares** [cuelang/cue#446] [General]
As you know, `cue` already has some builtin pkgs for different purposes, for example, you can use pkg `strings` to manipulate strings, that's very cool. 

Now if have a way/framework to develop customized or third pkg for `cue` for specific purposes, for example, in order to manage `Openstack` resources, and implement the action of manipulating each resource as a function that included in pkg `openstack`, then others can import them where they want, use them like standard pkg.

As talked with @myitcv in Slack thread , there could have three kinds of pkg:
> 1). The standard library
> 2). A curated set of well-maintained community packages
> 3). Everything else

There 3) could be promoted to 2), also 2) could be promoted to 1), pretty like Terraform Provider, everyone can follow its framework to develop customer Provider, then maintained by community or merged into official.

Welcome everyone who's interested this to share your opinions, suggestions or best practices here.

**Comment by mpvl** [comment]
The new evaluator makes it much easier to plug in additional packages programmatically, paving the way to make this possible.

The hardest part is good plugin architecture.

This could come in the form of packages, but it could also be in the form of a Go, WASM, JavaScript or what have you bridge of some sorts.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/446.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
